### PR TITLE
CSHRCP-925 - Added new endpoint to create or update a vacancy.

### PR DESCRIPTION
**Pull request by seamussingh**
**Please note the following:**

1. Compilation of this is dependent on being able to build and deploy the following projects (I don't have valid credentials for wherever Artifactory is now as the ones for bintray don't seem to work anymore for me):
	1. cshr-parent
	1. cshr-common 
1. The validation on Vacancy.atsVendorIdentifier has only been applied to the new /vacancy/save endpoint.  Further work will be required to implement this check on the POST /vacancy *create* and PUT /vacancy/{id} *update* endpoints. 
1. The newly introduced generic CSHRServiceStatus object has only be applied to the new endpoint and its exceptions. Further work will be required to use this through the rest of the code.
1. Unit tests for the new endpoint in VacancyControllerTest are outstanding.
1. In the Liquibase changelog.xml and testchangelog.xml files I tried to execute two sql statements as part of the changeset 27 and 28. The first one in Changeset 27 was: <sql>insert into atsvendor (clientidentifier, name) values('WCN0001', 'WCN V9 ATS';)</sql> and the one for Changeset 28 was <sql>update vacancies set atsvendoridentifier = 'WCN0001', version = 0;</sql>.  Although the log files recorded that Changeset 27 and 28 ran successfully and the sql statements used were logged there was no row of data in the atsvendor table.  There were no errors recorded and I have yet to figure out what the issue is. Therefore after deployment these two sql statements need to be run in whatever database is being used.  The value for atsvendor.clientIdentifier column can be anything right now but the new endpoint is to be used there must be a row in this table and the existing vacancies in the database should have this value set on the vacancies.atsvendoridentifer column.  The actual value to be used for WCN V9 will be determined a little later.
1. The combination of vacancies.identifier *job ref* and vacancies.atsvendoridentifer *ats vendor identifier* must be unique for each vacancy.

### CODE REVIEW

File: vcm/src/main/java/uk/gov/cshr/vcm/controller/VacancyController.java [hunk 1]
```diff
@@ -93,4 +105,48 @@
...
+     * that is recognised by CSHR.  If the identifier is not valid an InvalidApplicantTrackingSystemException will be thrown.
+     *
+     * The method will check if the vacancy already exists (based on Vacancy.identifier and Vacancy.atsVendorIdentifier).
+     * If the vacancy does not exist a new one will be created otherwise the existing one will be updated.
+     *
+     * @param vacancyToSave vacancy supplied to be saved in the data store
+     * @return Response include HttpStatus and instance of CSHRServiceStatus with information about whether the vacancye was created or updated.
+     */
+    @RequestMapping(method = RequestMethod.POST, value = "/save")
+    @ApiOperation(value = "Find all vacancies with support for pagination", nickname = "save")
```
Comment (reachsaurabh) [inline vcm/src/main/java/uk/gov/cshr/vcm/controller/VacancyController.java]: API operation value is incorrect

File: vcm/src/main/resources/application.yaml [hunk 2]
```diff
@@ -1,8 +1,8 @@
 spring:
   datasource:
     url: jdbc:postgresql://localhost:5432/cshr
-    username: postgres
-    password: qwerty
+    username: seamus
+    password: beast
```
[the remark this answers is not in the document]
Comment (reachsaurabh) [reply vcm/src/main/resources/application.yaml]: @RequestMapping(method = RequestMethod.PUT, value = "/{vacancyId}")
    @ApiOperation(value = "Update a Vacancy", nickname = "update")
    public ResponseEntity<Vacancy> update(@PathVariable Long vacancyId, @RequestBody Vacancy vacancyUpdate) {

This method seems redundant.

**Comment by reachsaurabh** [comment]
@ravigeda  please build the projects in order as per Seamus comments locally and use the feature branches itself for now to do the first round of testing . As their is some dependency on JK for this to get on the server , and we don't want to distract JK at this time , let me or Gordon know if you hit any issues . We will aim to get this on server ASAP .FYI -> @seamussingh @gmackay @jok-valtech Thanks.

File: vcm/src/main/java/uk/gov/cshr/vcm/controller/VacancyController.java [hunk 3]
```diff
@@ -93,4 +105,48 @@
...
+    @ApiResponses(value = {
+            @ApiResponse(code = 200, message = "Request to save vacancy was successful.", response = CSHRServiceStatus.class),
+            @ApiResponse(code = 422, message = "An unknown Applicant Tracking System Vendor was supplied.", response = CSHRServiceStatus.class)})
+    public ResponseEntity<CSHRServiceStatus> save(@RequestBody Vacancy vacancyToSave) {
+        applicantTrackingSystemService.validateClientIdentifier(vacancyToSave.getAtsVendorIdentifier());
+
+        List<Vacancy> vacancies = vacancyRepository.findVacancy(vacancyToSave.getIdentifier(), vacancyToSave.getAtsVendorIdentifier());
```
[the remark this answers is not in the document]
Comment (seamussingh) [reply vcm/src/main/java/uk/gov/cshr/vcm/controller/VacancyController.java]: No, but that is how JPA Queries work - they return List

File: vcm/src/main/java/uk/gov/cshr/vcm/controller/VacancyController.java (hunk 1 above)
Comment (seamussingh) [reply vcm/src/main/java/uk/gov/cshr/vcm/controller/VacancyController.java]: Doh, will correct

File: vcm/src/main/java/uk/gov/cshr/vcm/controller/VacancyController.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (reachsaurabh) [reply vcm/src/main/java/uk/gov/cshr/vcm/controller/VacancyController.java]: That's fine for now .. as discussed , I was looking a custom implement of a return type Optional<T> findBy if possible .

**Pull request closed without merging**